[MLIR] Determine contiguousness of memrefs with dynamic dimensions

momchil-velikov opened the pull request:
This patch enhances `MemRefType::areTrailingDimsContiguous` to also handle memrefs with dynamic dimensions.

The implementation itself is based on a new member function `MemRefType::getMaxCollapsableTrailingDims` that return the maximum number of trailing dimensions that can be collapsed - trivially all dimensions for memrefs with identity layout, or by examining the memref strides stopping at discontiguous or statically unknown strides.

(see also https://github.com/llvm/llvm-project/pull/142422)

momchil-velikov committed:
[MLIR] Determine contiguousness of memrefs with a dynamic dimensions

This patch enhances `MemRefType::areTrailingDimsContiguous`
to also handle memrefs with dynamic dimensions.

The implementation itself is based on a new member function
`MemRefType::getMaxCollapsableTrailingDims` that return
the maximum number of trailing dimensions that can be
collapsed - trivially all dimensions for memrefs with identity
layout, or by examining the memref strides stopping at discontguous
or statically unknown strides.

momchil-velikov committed:
[fixup] Fix an assertion

`computeStrides` does not acccess the first element of `sizes`

newling reviewed (commented):
Thanks, this looks good! I like the overall design, especially the new c++ testing approach. My comments are all minor.

momchil-velikov commented on mlir/include/mlir/IR/BuiltinTypes.td:
```diff
@@ -838,6 +838,20 @@ def Builtin_MemRef : Builtin_Type<"MemRef", "memref", [
     ///
     bool areTrailingDimsContiguous(int64_t n);
 
+    /// Return the maximum number of trailing dimensions that can be
```
Now I think this member function better be named `getMaxContiguousTrailingDims`.

As for the range of return values, my impression is where we talk about collapsing we talk about
collapsing `d0, d1, d2, ..., dn-1` dimensions into a single new one, as opposed to collapsing `d1, d2, ..., dn-1` into `d0`, thus range `[0, n]` seems most appropriate.

momchil-velikov replied:
Examples rewritten.

momchil-velikov committed:
[fixup] Misc NFC changes

 - rename `getMaxCollapsabelTrailingDims` to `getMaxContiguousTrailingDims`
 - new set of examples
 - remove redundant call to `isIdentify()`
 - make sure a variable type is visible on the declaration line
 - some micro-optimisation

momchil-velikov commented on mlir/lib/IR/BuiltinTypes.cpp:
```diff
@@ -646,35 +646,40 @@ LogicalResult MemRefType::verify(function_ref<InFlightDiagnostic()> emitError,
 bool MemRefType::areTrailingDimsContiguous(int64_t n) {
-  if (!isLastDimUnitStride())
-    return false;
+  return getLayout().isIdentity() ||
```
Done.

newling commented on mlir/lib/IR/BuiltinTypes.cpp:
```diff
@@ -646,35 +646,40 @@ LogicalResult MemRefType::verify(function_ref<InFlightDiagnostic()> emitError,
+  // if each stride `si` is the product of the dimensions `di+1, ..., dn-1`,
+  // for `i` in `[k, n-1]`.
+  int64_t dimProduct = 1;
+  for (int64_t i = n - 1; i >= 0; --i) {
```
`memref<1x4xi8, strided<[x,1]>>` is fully collapsible for any `x`. Basically, the stride in a unit dimension carries no information as there is no 'next element'. One way I think about it is to consider iterating through the elements. 
 
 ```
 for d0 in range(1):
   for d1 in range(4):
     element = vals[d0*x + d1] 
 ```

Unit dims correspond to loops with loop count of 1, which can be removed. Can you please add some tests for unit dims? Also fine if you prefer postponing the unit dim case for a later PR

momchil-velikov replied:
oh, I see.

momchil-velikov replied:
Done.

momchil-velikov committed:
[fixup] Handle unit dimensions by ignoring the corresponding stride

newling commented on mlir/lib/IR/BuiltinTypes.cpp:
```diff
@@ -646,35 +646,43 @@ LogicalResult MemRefType::verify(function_ref<InFlightDiagnostic()> emitError,
 bool MemRefType::areTrailingDimsContiguous(int64_t n) {
-  if (!isLastDimUnitStride())
-    return false;
+  return getMaxContiguousTrailingDims() >= std::min(n, getRank());
```
```suggestion
  assert(n <= getRank() && "doesn't have this many trailing dims to check for contiguity");
  return getMaxContiguousTrailingDims() >= n;
```

Maybe? Trying to understand the logic for the case where n > getRank().

momchil-velikov replied:
Nope, no logic here, an assertion is more appropriate.

newling commented on mlir/include/mlir/Dialect/Utils/IndexingUtils.h:
```diff
@@ -40,7 +40,7 @@ class ArrayAttr;
 ///   `[s1 * ... * sn, s2 * ... * sn, ..., sn, 1]`.
 ///
-/// `sizes` elements are asserted to be non-negative.
+/// `sizes` elements `s1` to `sn` are asserted to be non-negative.
```
Is this relaxation of the assert needed in this PR?

momchil-velikov replied:
Yes, it's needed because the leading dimension can appear dynamic.

newling commented on mlir/include/mlir/IR/BuiltinTypes.td:
```diff
@@ -838,6 +838,26 @@ def Builtin_MemRef : Builtin_Type<"MemRef", "memref", [
     ///
     bool areTrailingDimsContiguous(int64_t n);
 
+    /// Return the maximum number of trailing dimensions that are
```
```suggestion
    /// Return the number of trailing dimensions that are
```

Maybe 'maximum' is redundant, and the method could be called `getNumContiguousTrailingDims()`

momchil-velikov replied:
Sounds good.

newling commented on mlir/lib/Dialect/Utils/IndexingUtils.cpp:
```diff
@@ -69,7 +69,8 @@ SmallVector<ExprType> delinearizeImpl(ExprType linearIndex,
 
 SmallVector<int64_t> mlir::computeSuffixProduct(ArrayRef<int64_t> sizes) {
-  assert(llvm::all_of(sizes, [](int64_t s) { return s >= 0; }) &&
+  assert((sizes.size() == 0 ||
```
```suggestion
  assert((!sizes.empty() ||
```

momchil-velikov replied:
Done

newling reviewed (commented):
Thank for for adding the stride = 1 cases, looks good to me! I've added a few minor comments.

momchil-velikov committed:
[fixup] Address misc review comments

newling reviewed (approved):
LGTM! Maybe allow others with more experience here than me some time to review

momchil-velikov committed:
[fixup] Fix a couple of FileCheck lines

ftynse commented on mlir/unittests/Dialect/MemRef/LayoutTest.cpp:
```diff
@@ -0,0 +1,210 @@
+  MLIRContext ctx;
+  OpBuilder b(&ctx);
+
+  const auto _ = ShapedType::kDynamic;
+  const auto f32 = b.getF32Type();
```
https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

In general, MLIR uses `auto` when the type is obvious from line-level context (cast, constructor) or too long to spell.

ftynse commented on mlir/unittests/Dialect/MemRef/LayoutTest.cpp:
We usually don't write such unit tests when there exists a way to check functionality via a pass. Individual unit tests create a lot of additional compiler build time. In this case, one could add a test pass that annotates the operation with an attribute containing the number of contiguous dimensions on the result type, for example.

Also, I see this is carried over from a pre-existing structure, but this doesn't belong to `unittest/Dialect/MemRef`. The memref type is in the builtin dialect, so the related tests should live in `unittest/IR`.

newling replied:
Perhaps a pass like this one which tests the another method of memref: 

https://github.com/llvm/llvm-project/blob/afd31339458967d3da9a04379f46f5e9ec1bb587/mlir/test/lib/Analysis/TestMemRefStrideCalculation.cpp#L55

https://github.com/llvm/llvm-project/blob/afd31339458967d3da9a04379f46f5e9ec1bb587/mlir/test/Dialect/Affine/memref-stride-calculation.mlir#L77

momchil-velikov replied:
I consider those `test-` passes an abuse of both the pass pipeline and the testing infrastructure.

Of course, if I get an unambiguous statement "**this PR is not going in with those unit tests**", I'll do it (with great displeasure).

momchil-velikov replied:
> I'll do it

Or rather, resurrect it, I had that pass earlier that did
```
/// Traverse AllocOp and compute the maximum number of contiguous trailing dimensions.
void TestMemRefContigCalculation::runOnOperation() {
  getOperation().walk([&](func::CallOp callOp) {
    auto memrefType = cast<MemRefType>(callOp.getResult(0).getType());
    memrefType.print(llvm::outs());

    int64_t n = memrefType.getMaxContiguousTrailingDims();
    llvm::outs() << ": " << n << " contiguous trailing dims(s)\n";
  });
  llvm::outs().flush();
}
```

newling replied:
(I'm fine with either)

banach-space replied:
Let me chime in briefly.

> We usually don't write such unit tests when there exists a way to check functionality via a pass.

Just to clarify: there isn’t a way to test `getNumContiguousTrailingDims` in isolation today. That said, we do have transformations that use it.

**Question 1:** Is it valuable to test `getNumContiguousTrailingDims` in isolation?
I think so - it checks a fairly fundamental property, and the test effectively documents what qualifies as a “contiguous memref slice.”

**Question 2:** How do we test it?
There are two main options: unit tests or a test pass (possibly with a TD op):
* Unit tests involve less boilerplate but increase compile times.
* A test pass avoids compile-time overhead but requires more setup. Runtime cost is unclear - though honestly, we’re hand-waving a bit here without measurements.

Personally, I don’t have a strong preference. I’ve tried to lay out the trade-offs, and I think either direction is valid. In cases like this, I’d suggest just sticking with the approach that’s already implemented.

@ftynse, I read your comment more as a suggestion than a hard requirement - please let us know if we misread that. (As I said, I’m happy with either option.)

momchil-velikov committed:
[fixup] Add some negative tests with dynamic dimensions

momchil-velikov committed:
[fixip] Move a test and spell out some auto types

newling commented:
Just a drive-by comment - when I started contributing here, a reviewer kindly pointed out to me in this https://github.com/llvm/llvm-project/pull/135096#issuecomment-2797902624 that force pushing isn't recommended https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

banach-space commented on mlir/unittests/IR/MemrefLayoutTest.cpp:
A few high-level comments on test organization and naming:

* `maxContigDim` tests `getNumContiguousTrailingDims`, `contigTrailingDim` tests `areTrailingDimsContiguous`, and `identityMaps` tests both - is there an intended pattern here?
* The actual "complex" logic seems to reside in `getNumContiguousTrailingDims`. Would it make sense to focus the tests more narrowly on that hook?
* Could the test names (`maxContigDim`,` contigTrailingDim`, `identityMaps`) be made more descriptive? Alternatively, adding comments explaining what each test is validating would help - it’s not immediately clear to me.
* I’d suggest renaming `identityMaps` to something like `noStrides` or `defaultStrides`, since the other tests explicitly include strides. The natural split seems to be "with strides" vs "without strides".

banach-space commented on mlir/unittests/IR/MemrefLayoutTest.cpp:
```diff
@@ -0,0 +1,210 @@
+    return StridedLayoutAttr::get(&ctx, 0, s);
+  };
+
+  // memref<2x2x2xf32, strided<[4,2,1]>
+  auto m1 = MemRefType::get({2, 2, 2}, f32, strided({4, 2, 1}));
+  EXPECT_EQ(m1.getNumContiguousTrailingDims(), 3);
+
+  // memref<2x2x2xf32, strided<[8,2,1]>
+  auto m2 = MemRefType::get({2, 2, 2}, f32, strided({8, 2, 1}));
+  EXPECT_EQ(m2.getNumContiguousTrailingDims(), 2);
+
+  // memref<2x2x2xf32, strided<[8,4,1]>
+  auto m3 = MemRefType::get({2, 2, 2}, f32, strided({8, 4, 1}));
+  EXPECT_EQ(m3.getNumContiguousTrailingDims(), 1);
+
+  // memref<2x2x2xf32, strided<[8,4,2]>
+  auto m4 = MemRefType::get({2, 2, 2}, f32, strided({8, 4, 2}));
+  EXPECT_EQ(m4.getNumContiguousTrailingDims(), 0);
```
This is a nice variation of the initial example, `memref<2x2x2xf32, strided<[4,2,1]>`, where modifying the strides alters the number of contiguous dims. However, the examples that follow (with dynamic shapes) feel arbitrary. Could you group them somehow, either with block comments or by splitting into multiple test functions.

momchil-velikov replied:
They are not arbitrary, they are systematically created. Splitting into multiple test functions sounds  arbitrary and unnecessary.

banach-space commented on mlir/unittests/IR/MemrefLayoutTest.cpp:
```diff
@@ -0,0 +1,210 @@
+  EXPECT_EQ(m18.getNumContiguousTrailingDims(), 3);
+}
+
+TEST(MemRefLayout, contigTrailingDim) {
```
Given that this only tests `areTrailingDimsContiguous` and `areTrailingDimsContiguous`  simply delegates to `getNumContiguousTrailingDims`, do we really need this?

momchil-velikov replied:
Yes.

momchil-velikov replied:
The function was rewritten entirely, as such it requires a test case.

I've gone down to two invocations testing the two possible return values of the function.

banach-space commented on mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:
```diff
@@ -188,9 +188,31 @@ func.func @transfer_read_leading_dynamic_dims(
 // -----
 
-// One of the dims to be flattened is dynamic - not supported ATM.
+// The vector could be a non-contiguous slice of the input
+// memref.
```
Could you explain _why_ "could"?

momchil-velikov replied:
Are you objecting about the word "could"? Perhaps you prefer "is" ?

banach-space commented on mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:
```diff
@@ -188,9 +188,31 @@ func.func @transfer_read_leading_dynamic_dims(
+
+// -----
+
+// Can flatten the righmost dynamic dimension
```
Why? Because the corresponding vector dim is 1? Could you expand the comment?

momchil-velikov replied:
Some more context, the comment and the test were originally

```
// One of the dims to be flattened is dynamic - not supported ATM.

func.func @negative_transfer_read_dynamic_dim_to_flatten(
...
```

With this patch the test is no longer negative and the comment reflects that
at the same level of detail  as the original comment.

banach-space commented on mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:
```diff
@@ -551,6 +620,8 @@ func.func @negative_out_of_bound_transfer_read(
 
 // -----
 
+// Can flatten the righmost dynamic dimension
```
"Can"? But it's not flattened?

momchil-velikov replied:
That comment is accidental, no idea where is came from.

banach-space reviewed (commented):
LG % some suggestion re structuring the unit tests.

I suggest leaving ~24hrs for Alex to comment whether he minds you leaving the unit tests.

momchil-velikov commented:
Sometime a rebase is not a matter of preference, but unavoidable, for example if you have stacked PRs and add a fixup to a dependent branch (not that case here).

momchil-velikov committed:
[fixup] Misc unimportant changes

momchil-velikov commented on mlir/test/Dialect/Vector/vector-transfer-flatten.mlir:
```diff
@@ -190,7 +190,7 @@ func.func @transfer_read_leading_dynamic_dims(
 
 // One of the dims to be flattened is dynamic - not supported ATM.
 
-func.func @negative_transfer_read_dynamic_dim_to_flatten(
```
Done

momchil-velikov committed:
[fixup] Address review comments

momchil-velikov committed:
[fixup] Reduce the number of test cases

banach-space reviewed (approved):
Thank you for the updates, the latest revision addresses my comment, LGTM!

From what I can tell, all comments have been addressed, but please wait till tomorrow before merging. Just in case there's more comments (also being mindful of the time difference between contributors).

Thanks for seeing this through!
